config/cloudinit/datasource/datasource.go

```diff
@@ -30,12 +32,18 @@ type Datasource interface {
 	Finish() error
 }
 
+// TODO: move to netconf/types.go ?
+// see https://ahmetalpbalkan.com/blog/comparison-of-instance-metadata-services/
 type Metadata struct {
-	PublicIPv4    net.IP
-	PublicIPv6    net.IP
-	PrivateIPv4   net.IP
-	PrivateIPv6   net.IP
 	Hostname      string
 	SSHPublicKeys map[string]string
-	NetworkConfig interface{}
+	NetworkConfig netconf.NetworkConfig
+
+	// probably unused, but its in the initialize.env code
+	// TODO: work out if there's any reason to keep it.
+	// Lets see if anyone notices when its not set.
+	GoingPublicIPv4  net.IP
```

Comment:
There was functionality to replace $public_ipv4 and similar with the IPs.  We should keep this and make it more useful.  This is a really common thing in AWS and other places in that you want the IP of the server but without requiring scripting (like just embed it as a environment variable to a service).